ibtk/src/lagrangian/FEProjector.cpp

```diff
@@ -357,6 +365,164 @@ FEProjector::buildLumpedL2ProjectionSolver(const std::string& system_name)
+
+        // Build solver components.
+        std::unique_ptr<PetscLinearSolver<double> > solver(new PetscLinearSolver<double>(comm));
+
+        std::unique_ptr<PetscMatrix<double> > M_mat(new PetscMatrix<double>(comm));
+        M_mat->attach_dof_map(dof_map);
+        M_mat->init();
+        MatSetOption(M_mat->mat(), MAT_IGNORE_ZERO_ENTRIES, PETSC_TRUE);
+        MatSetOption(M_mat->mat(), MAT_SPD, PETSC_TRUE);
+        MatSetOption(M_mat->mat(), MAT_SYMMETRY_ETERNAL, PETSC_TRUE);
+
+        // Loop over the mesh to construct the system matrix.
+        DenseMatrix<double> M_e;
+        DenseVector<double> Pi_phi_e;
+        std::vector<libMesh::dof_id_type> dof_id_scratch;
+        const MeshBase::const_element_iterator el_begin = mesh.active_local_elements_begin();
+        const MeshBase::const_element_iterator el_end = mesh.active_local_elements_end();
+        for (MeshBase::const_element_iterator el_it = el_begin; el_it != el_end; ++el_it)
+        {
+            const Elem* const elem = *el_it;
+            fe->reinit(elem);
+            const auto& dof_indices = dof_map_cache.dof_indices(elem);
+            for (unsigned int var_n = 0; var_n < dof_map.n_variables(); ++var_n)
+            {
+                const auto& dof_indices_var = dof_indices[var_n];
+                const auto n_basis = static_cast<unsigned int>(dof_indices_var.size());
+                M_e.resize(n_basis, n_basis);
+                Pi_phi_e.resize(n_basis);
+                const unsigned int n_qp = qrule->n_points();
+
+                const double vol_e = elem->volume();
+                for (unsigned int i = 0; i < n_basis; ++i)
+                {
+                    for (unsigned int qp = 0; qp < n_qp; ++qp)
+                    {
+                        Pi_phi_e(i) += phi[i][qp] * JxW[qp] / vol_e;
+                    }
+                }
+
+                for (unsigned int i = 0; i < n_basis; ++i)
+                {
+                    for (unsigned int j = 0; j < n_basis; ++j)
+                    {
+                        for (unsigned int qp = 0; qp < n_qp; ++qp)
+                        {
+                            M_e(i, j) += ((phi[i][qp] * phi[j][qp]) +
+                                          epsilon * (phi[i][qp] - Pi_phi_e(i)) * (phi[j][qp] - Pi_phi_e(j))) *
+                                         JxW[qp];
+                        }
+                    }
+                }
+
+                copy_dof_ids_to_vector(var_n, dof_indices, dof_id_scratch);
+                dof_map.constrain_element_matrix(M_e,
+                                                 dof_id_scratch,
+                                                 /*assymetric_constraint_rows*/ false);
+                prune_roundoff_entries(M_e);
+                M_mat->add_matrix(M_e, dof_id_scratch);
+            }
+        }
+
+        // Flush assemble the matrix.
+        M_mat->close();
+
+        // Reset values at Dirichlet boundaries.
```

Comment:
I am still pretty sure that this code block (which we have in multiple places) doesn't actually do anything since we are now applying constraints correctly. Do we have any examples where we actually use it?

Reply:
I am not sure that we do.

Reply:
I don't know if we always set up proper Dirichlet constraints from these tags. (In the past, at least, I know that we did _not_.)

Reply:
That's something else that we should add to the set of constraints. Let's just keep the status quo for now.